### DISCUSSION TITLE
passing variables to scripts in pollvars?

### DISCUSSION THREAD

**Discussion by ghistes** [elkowar/eww#300] [Q&A]
Hi,

I would like to implement a clock that displays the time in different formats. I want to be able to cycle through the different formats by clicking on it.

So I thought a construction like this would work:

    (defvar clock_state "time_only")
    (defpoll time :interval "10s"
       "/home/mh/test/eww/format_time.pl ${clock_state}")

So in effect you have a defvar that keeps track of the format that should be displayed and you pass that into a script that then formats the current time accordingly and you simply display this defpoll-variable.

However the above construction does not work ("error: Tried to reference variable `clock_state`, but we cannot access variables here") and because the clock_state variable is not used in a widget it is also not visible via "eww state" (is that a bug?).

So is there a way to make an external script aware of the internal state of eww?

Ideally I would like to have both a way to pass variables to external-scripts as well as maybe an "eww get <variable>" command, so that an external script could easily obtain such variables-values by querying the daemon.

**Comment by elkowar** [comment]
It's currently not possible to directly reference variables in script-vars, for several reasons. This _might_ change in the future, but it has a good bit of implied complexity with regards to state management.
`eww state` only shows the variables that are currently referenced, that is correct. The reason for this is that poll and tail vars actually don't get computed while they're not used, which is a performance optimization. `eww state --all` can show even unused variables. In case of listen and poll vars, they still don't get updated while they're not referenced in any widgets. simple `defvar` variables that are manually updated however should get included and handled properly just fine with `eww state --all`. 

A `eww get` command or something along those lines might be considered in the future, feel free to file a feature request 😄 

There would be multiple other ways to work around this currently: you could either have multiple script-vars for different time formats and cycle through using a different var - this is kinda ugly

additionally, you can currently of course just implement the cycling outside of eww, and just have it be part of your format_time.pl script itself.

**Reply by trueNAHO** [reply]
FYI, `eww get` is a valid subcommand in version `0.4.0`:

```txt
$ eww --help
SUBCOMMANDS:
    get          Get the value of a variable if defined
```

**Comment by Wallboy** [comment]
Hi,

Regarding variables not created/ran until they are referenced, if you need to create a variable (defvar, defpoll, deflisten) without the need to visibly reference it, you can do something like the following:

```
(defpoll background_script :interval "10s" `scripts/do_something.sh`)
 
(defwidget dummy []
    (box :visible false :width -1 :height -1
        {background_script}
    )
)
```

Then place this dummy widget anywhere as a child within your window. For example:

```
(defwindow my_window
           :stacking "bg"
           :geometry (geometry :x "0px"
                               :y "0px"
                               :anchor "top left")
           :windowtype "desktop"
  (main_container)
)

(defwidget main_container []
    (dummy)
    (other_widgets...)
)
```
